### PR TITLE
Safe strings

### PR DESCRIPTION
**Pull request by cedlemo**
I wanted to use your package but opam was not able to install it with OCaml 4.06 which has the option safe-strings enable by default.

I don't know if it is the best way to fix that, but I was able to build shcaml with the following changes.

Regards.

### CODE REVIEW

File: lib/channel.ml [hunk 1]
```diff
@@ -289,7 +289,7 @@ let string_of_channel c =
   let buf     = Buffer.create bufsize in
   let str     = String.make bufsize '\000' in
   let rec loop c =
-    match Pervasives.input c str 0 bufsize with
+    match Pervasives.input c Bytes.(of_string str) 0 bufsize with
```
Comment (Armael) [inline lib/channel.ml:292]: This does not actually work: it creates a writable copy of `str` (of type `bytes`), writes the input in it, and immediately throws it away. I think you should instead create `str` as bytes above (using `Bytes.make`), and use `Buffer.add_subbytes` instead of `Buffer.add_substring`.

File: lib/iVar.ml [hunk 2]
```diff
@@ -33,7 +33,7 @@ let write rep v =
   kill rep;
   Unix.close rep.read_end;
   let buf = Marshal.to_string v [Marshal.Closures] in
-  ignore @@ Unix.write rep.write_end buf 0 (String.length buf);
+  ignore @@ Unix.write rep.write_end Bytes.(of_string buf) 0 (String.length buf);
```
Comment (Armael) [inline lib/iVar.ml:36]: You can use `Marshal.to_bytes` directly to avoid copying.

**Comment by tov** [comment]
This seems fine and good to me. @Armael?

File: lib/line.ml [hunk 3]
```diff
@@ -97,7 +97,7 @@ module Key_value = struct
     try Some (float_of_string (value line)) with Failure _ -> None
 
   let as_bool line =
-    match String.lowercase (value line) with
+    match String.lowercase_ascii (value line) with
```
Comment (Armael) [inline lib/line.ml:100]: This should go in a separate PR (`String.lowercase_ascii` has been introduced in ocaml 4.03, and we support ocaml >= 4.02, so as-is this change would break compatibility). A way of dealing with that issue is using https://github.com/thierry-martinez/stdcompat -- could you open a separate PR with that change + adding Stdcompat?

File: lib/reader.ml [hunk 4]
```diff
@@ -66,8 +66,8 @@ let rec make =
       let result = { content  = String.make n '\000';
                      before   = "";
                      after    = String.make m '\000'; } in
-      Pervasives.really_input c result.content 0 n;
-      ignore @@ Pervasives.input c result.after 0 n;
+      Pervasives.really_input c Bytes.(of_string result.content) 0 n;
+      ignore @@ Pervasives.input c Bytes.(of_string result.after) 0 n;
```
Comment (Armael) [inline lib/reader.ml:70]: Same here: I don't think this works because it immediately throws away the `bytes` buffer created. You should allocate the buffers as bytes, do the input, then convert them to strings when creating the result. I think in this context we can use `Bytes.unsafe_to_string` to avoid copying, since we know this function is the unique owner of the buffers.

**Review by Armael — CHANGES_REQUESTED** [review]
Thanks for the PR, this was needed indeed! I added some comments; I'll be happy to merge this when they're addressed.

**Comment by cedlemo** [comment]
#5
